attrs_strict/_type_validation.py

```diff
@@ -78,6 +80,70 @@ def _validate_elements(attribute, value, expected_type):
         _handle_tuple(attribute, value, expected_type)
     elif base_type == typing.Union:
         _handle_union(attribute, value, expected_type)
+    elif base_type in SimilarTypes.Callable:
+        _handle_callable(attribute, value, expected_type)
+
+
+def _get_base_type(type_):
+    if hasattr(type_, "__origin__") and type_.__origin__ is not None:
+        base_type = type_.__origin__
+    elif is_newtype(type_):
+        base_type = type_.__supertype__
+    else:
+        base_type = type_
+
+    return base_type
+
+
+def _type_matching(actual, expected):
+    actual = _get_base_type(actual) if is_newtype(actual) else actual
+    expected = _get_base_type(expected) if is_newtype(expected) else expected
+
+    if expected == actual or expected == typing.Any:
+        return True
```

Comment:
This function is a recursive call, where it relies on the return value's truthy-ness.

In the case of `Union` the types would match if `any` of the candidate-types match. Whereas for all other cases it is `all`. Due to the nature of `Union`, a mismatch in the recursive tree may or may not contribute to a `CallableError`. Hence I think it would be a bit misleading to raise one in this function.

Regarding shortcircuiting, It would still do it as soon as it can. Both [any](https://docs.python.org/3/library/functions.html#any) and [all](https://docs.python.org/3/library/functions.html#all) functions support shortcircuiting. The caller will also shortcircuit the result on the first mismatch - https://github.com/bloomberg/attrs-strict/blob/a16be1a05950a318b150a3d7cbcb5525b65d4704/attrs_strict/_type_validation.py#L140-L144

Reply:
Ok I think that makes sense, the fact that you couldn't handle the type in a level below is not necessary an error.